### PR TITLE
Create coarse triangulations for FullyDistributed

### PR DESCRIPTION
**Pull request by necioglu**
Create the coarse triangulations needed for global coarsening multigrid manually in grid class.

### CODE REVIEW

File: include/exadg/grid/grid.cpp [hunk 1]
```diff
@@ -160,10 +160,47 @@ Grid<dim>::create_triangulation(
...
+      while(refine_local != std::vector<unsigned int>(refine_local.size(), 0))
+      {
+        for(size_t material_id = 0; material_id < refine_local.size(); material_id++)
+        {
+          if(refine_local[material_id] > 0)
+          {
+            refine_local[material_id]--;
+          }
+        }
+        lambda_create_coarse_levels(0, refine_local);
```
Comment (peterrum) [inline include/exadg/grid/grid.cpp:202]: I don't understand the logic here. Could you explain this?

Comment (necioglu) [reply include/exadg/grid/grid.cpp:202]: We use the `do_create_triangulation` repeatedly to create the coarse triangulations needed. First we "undo" the global refinements, i.e., we reduce the global refinement by one each time until there is no global refinement and save the triangulation to the `coarse_triangulations` vector. Afterwards we "undo" the `local_refinements` by checking if the `refine_local` vector is all zeros, i.e., there are no local refinements left. If there are local refinements, we reduce each entity by one,build the triangulation, save it and continue reducing until there is none.

File: include/exadg/grid/grid.cpp [hunk 2]
```diff
@@ -160,10 +160,47 @@ Grid<dim>::create_triangulation(
...
 
-      AssertThrow(false, ExcNotImplemented());
+      // start with the finest level
+      std::vector<unsigned int> refine_local = vector_local_refinements;
+
+      // undo global refinements
+      for(int it = global_refinements; it >= 0; --it)
+      {
+        lambda_create_coarse_levels(it, refine_local);
+      }
```
Comment (peterrum) [inline include/exadg/grid/grid.cpp:190]:
```suggestion
      std::vector<unsigned int> refine_local = vector_local_refinements;
      for(auto & i : refine_local)
        i += global_refinements;
```

File: include/exadg/grid/grid.cpp [hunk 3]
```diff
@@ -160,10 +160,47 @@ Grid<dim>::create_triangulation(
...
+
+      // undo local refinements
+      while(refine_local != std::vector<unsigned int>(refine_local.size(), 0))
```
Comment (peterrum) [inline include/exadg/grid/grid.cpp:193]: Please use `std::max(...)!=0`.

File: include/exadg/grid/grid.cpp (hunk 1 above)
Comment (peterrum) [reply include/exadg/grid/grid.cpp:202]: Makes sense. I have suggested a way to simplify the code a bit.

File: include/exadg/grid/grid.cpp (hunk 2 above)
Comment (necioglu) [reply include/exadg/grid/grid.cpp:190]: I don't see how this would do the same thing as the original code.

File: include/exadg/grid/grid.cpp (hunk 3 above)
Comment (necioglu) [reply include/exadg/grid/grid.cpp]: done!

File: include/exadg/grid/grid.cpp (hunk 2 above)
Comment (necioglu) [reply include/exadg/grid/grid.cpp:190]: Yes that makes sense if we always have a local refinement, but what happens if the `refine_local` vector is empty? Then the `global_refinements` are also disregarded, and no `coarse_triangulations` are created.

Comment (peterrum) [reply include/exadg/grid/grid.cpp:190]: Can that be empty? If yes let's keep it like this!

Comment (necioglu) [reply include/exadg/grid/grid.cpp:190]: Yes, it is the default case.

https://github.com/exadg/exadg/blob/7edf07c5a30582accdd72ce2af833caa51382107/include/exadg/grid/grid.h#L68-L73

**Review by bergbauer — APPROVED** [review]
LGTM!

File: include/exadg/grid/grid.cpp [hunk 4]
```diff
@@ -160,10 +160,50 @@ Grid<dim>::create_triangulation(
...
-      // TODO
-      // construct all the coarser triangulations
+      // resize the empty coarse triangulations vector
+      coarse_triangulations = std::vector<std::shared_ptr<dealii::Triangulation<dim> const>>(
+        triangulation->n_global_levels());
+
+      unsigned int coarse_level = triangulation->n_global_levels() - 1;
+
+      // lambda function for creating and placing the coarse triangulations
+      auto const lambda_create_coarse_levels = [&](unsigned int              global_refinement,
```
Comment (nfehn) [inline include/exadg/grid/grid.cpp:170]: the name should be `refine_global` for consistency with the second argument `refine_local`.

File: include/exadg/grid/grid.cpp [hunk 5]
```diff
@@ -160,10 +160,50 @@ Grid<dim>::create_triangulation(
     {
       // we need to generate the coarse triangulations explicitly
 
-      // TODO
-      // construct all the coarser triangulations
+      // resize the empty coarse triangulations vector
+      coarse_triangulations = std::vector<std::shared_ptr<dealii::Triangulation<dim> const>>(
+        triangulation->n_global_levels());
+
+      unsigned int coarse_level = triangulation->n_global_levels() - 1;
```
Comment (nfehn) [inline include/exadg/grid/grid.cpp:167]: The name `coarse_level` is misleading, because the lambda `create_coarse_triangulation` refers to the coarsest level. Hence, there is a mismatch in notation. Just use `level` instead, and add a comment that you start, by convention, with the highest level.

File: include/exadg/grid/grid.cpp (hunk 4 above)
Comment (nfehn) [inline include/exadg/grid/grid.cpp:170]: The name `lambda_create_coarse_levels` does also not fit very well, since the lambda function creates the triangulation just for one level, not for all. The current name reads like the function creates all coarse levels at once. I suggest `lambda_create_level_triangulation` for consistency with the other names.

File: include/exadg/grid/grid.cpp [hunk 6]
```diff
@@ -160,10 +160,50 @@ Grid<dim>::create_triangulation(
...
+
+      // lambda function for creating and placing the coarse triangulations
+      auto const lambda_create_coarse_levels = [&](unsigned int              global_refinement,
+                                                   std::vector<unsigned int> refine_local) {
+        auto new_tria = std::make_shared<dealii::parallel::fullydistributed::Triangulation<dim>>(
```
Comment (nfehn) [inline include/exadg/grid/grid.cpp:172]: I would then use `level_tria` here, because `new` is very general and can mean a lot.

File: include/exadg/grid/grid.cpp [hunk 7]
```diff
@@ -144,6 +142,8 @@ Grid<dim>::create_triangulation(
     // coarse grid triangulations
     if(data.triangulation_type == TriangulationType::Serial)
     {
+      AssertThrow(triangulation->all_reference_cells_are_hyper_cube(), dealii::ExcNotImplemented());
```
Comment (nfehn) [inline include/exadg/grid/grid.cpp:145]: In my opinion, this needs some more explanations. As a new user/developer of ExaDG, I can deduce that Serial requires hypercubes, but I would like to know the reason, e.g. that this is a current limitation of deal.II. Why does simplex not work here? I think this should be mentioned here. Imagine deal.II becomes more powerful. How should a developer of ExaDG know in 1-2 years why we have this assert in the code and how should he decide whether it can be removed or not. A developer might remove it and then realize it works, but he will then ask himself whether this is just luck and whether it will work in all cases ...

Comment (necioglu) [reply include/exadg/grid/grid.cpp:145]: That is why I used `dealii::ExcNotImplemented()` and not `ExaDG::ExcNotImplemented()`. I thought this made clear, that currently this is not allowed, because of dealii and not because of ExaDG. However, I agree that adding more explanation would be helpful.

File: include/exadg/grid/grid.cpp (hunk 6 above)
Comment (necioglu) [reply include/exadg/grid/grid.cpp:172]: I agree. I wanted to keep our naming similar to dealii. See the equivalent function in dealii using the variable name `new_tria`: https://github.com/dealii/dealii/blob/07b5a85db9d97aac577b56873c6c34ff4f3bf73a/include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h#L2439-L2451

File: include/exadg/grid/grid.cpp (hunk 4 above)
Comment (necioglu) [reply include/exadg/grid/grid.cpp]: changed in fe49537a6dd436ad6b49130feb488cd584e5d308.

Comment (necioglu) [reply include/exadg/grid/grid.cpp]: changed in fe49537a6dd436ad6b49130feb488cd584e5d308.

File: include/exadg/grid/grid.cpp (hunk 6 above)
Comment (necioglu) [reply include/exadg/grid/grid.cpp]: changed in fe49537a6dd436ad6b49130feb488cd584e5d308.

File: include/exadg/grid/grid.cpp (hunk 7 above)
Comment (necioglu) [reply include/exadg/grid/grid.cpp]: changed in fe49537a6dd436ad6b49130feb488cd584e5d308.

File: include/exadg/grid/grid.cpp [hunk 8]
```diff
@@ -160,10 +164,51 @@ Grid<dim>::create_triangulation(
...
+
+        level--;
+      };
 
-      AssertThrow(false, ExcNotImplemented());
+      // start with the finest level
+      std::vector<unsigned int> refine_local = vector_local_refinements;
+
+      // undo global refinements
+      for(int it = global_refinements; it >= 0; --it)
```
Comment (nfehn) [inline include/exadg/grid/grid.cpp:192]: a last suggestion for renaming would be to use `refine_global` here instead of `it`.

**Comment by nfehn** [comment]
I made one more comment. Apart from that, it looks very good from my side!

File: include/exadg/grid/grid.cpp (hunk 8 above)
Comment (necioglu) [reply include/exadg/grid/grid.cpp]: done!

File: include/exadg/grid/grid.cpp [hunk 9]
```diff
@@ -160,10 +164,51 @@ Grid<dim>::create_triangulation(
...
+                                                         std::vector<unsigned int> refine_local) {
+        auto level_tria = std::make_shared<dealii::parallel::fullydistributed::Triangulation<dim>>(
+          triangulation->get_communicator());
+
+        do_create_triangulation(
+          level_tria, data, create_coarse_triangulation, refine_global, refine_local);
+
+        coarse_triangulations[level] = level_tria;
+
+        level--;
```
[the remark this answers is not in the document]
Comment (nfehn) [reply include/exadg/grid/grid.cpp:185]: wouldn't it be better if the lambda returned the coarse triangulation and filling the vector is done outside the lambda?

[the remark this answers is not in the document]
Comment (necioglu) [reply include/exadg/grid/grid.cpp]: done!

File: include/exadg/grid/grid.cpp [hunk 10]
```diff
@@ -160,10 +164,52 @@ Grid<dim>::create_triangulation(
...
 
-      // TODO
-      // construct all the coarser triangulations
+      // resize the empty coarse triangulations vector
+      coarse_triangulations = std::vector<std::shared_ptr<dealii::Triangulation<dim> const>>(
+        triangulation->n_global_levels());
+
+      // we start with the finest level
+      unsigned int              level        = triangulation->n_global_levels() - 1;
+      std::vector<unsigned int> refine_local = vector_local_refinements;
```
Comment (nfehn) [inline include/exadg/grid/grid.cpp:173]: Why do you define these variables here? Wouldn't it make more sense to shift these lines to line 187 below.

**Comment by nfehn** [comment]
Some of the checks were not successful. Overall, I think this PR can be merged with only some minor changes, so it would be good to make progress here and complete this PR.

File: include/exadg/grid/grid.cpp (hunk 10 above)
Comment (necioglu) [reply include/exadg/grid/grid.cpp]: done!

Comment (necioglu) [reply include/exadg/grid/grid.cpp]: I also moved lines 167-169.

**Comment by necioglu** [comment]
> Some of the checks were not successful. Overall, I think this PR can be merged with only some minor changes, so it would be good to make progress here and complete this PR.

I wasn't sure on what to do the about the failing checks. They didn't seemed like they were caused because of this PR, since the master is/was failing also. I rebased and pushed now.

**Comment by nfehn** [comment]
looks good from my side

File: applications/poisson/sine/application.h [hunk 11]
```diff
@@ -185,24 +185,24 @@ class Application : public ApplicationBase<dim, n_components, Number>
...
+      }
+      else
+      {
+        AssertThrow(false, ExcNotImplemented());
+      }
+    };
+
+    this->grid->create_triangulation(this->param.grid,
+                                     lambda_create_coarse_triangulation,
+                                     this->param.grid.n_refine_global);
```
Comment (nfehn) [inline applications/poisson/sine/application.h:205]: According to my current understanding, this change here is problematic in terms of the bug discussed in issue #313 because the part with Neumann boundary conditions below is done after this function call, which is not allowed, because the `coarse_grid_triangulations` for global-coarsening multigrid then have the wrong boundary conditions. So I conclude the bug has not been introduced by changes in #269, but PR #269 did not update this particular test case (poisson/sine) so that it can be used for global coarsening multigrid.

Comment (necioglu) [reply applications/poisson/sine/application.h:205]: I am looking into this!

Comment (nfehn) [reply applications/poisson/sine/application.h:205]: I moved the Neumann part into the lambda and it works. Im already working on it.

Comment (necioglu) [reply applications/poisson/sine/application.h:205]: Okay. Thanks!

Comment (nfehn) [reply applications/poisson/sine/application.h:205]: The problem is that - by the current design - we can not solve the problem with curvilinear mesh in poisson/sine correctly, see issue #319.
